File: c/TennisGame1.c
#include <stdlib.h>
#include <string.h>

#include "TennisGame.h"

struct TennisGame
{
    int m_score1;
    int m_score2;
    const char* player1Name;
    const char* player2Name;
    char score[18];
};

struct TennisGame* TennisGame_Create(const char* player1Name, const char* player2Name)
{
    struct TennisGame* newGame = malloc(sizeof(struct TennisGame));
    newGame->m_score1 = 0;
    newGame->m_score2 = 0;
    newGame->player1Name = player1Name;
    newGame->player2Name = player2Name;
    return newGame;
}

void TennisGame_WonPoint(struct TennisGame* game, const char* playerName)
{
    if (strcmp(playerName, "player1") == 0)
        game->m_score1 += 1;
    else
        game->m_score2 += 1;
}

const char* TennisGame_GetScore(struct TennisGame* game)
{
    game->score[0] = '\0';
    int tempScore = 0;
    if (game->m_score1 == game->m_score2)
    {
        switch (game->m_score1)
        {
        case 0:
            strcpy(game->score, "Love-All");
            break;
        case 1:
            strcpy(game->score, "Fifteen-All");
            break;
        case 2:
            strcpy(game->score, "Thirty-All");
            break;
        default:
            strcpy(game->score, "Deuce");
            break;

        }
    }
    else if (game->m_score1 >= 4 || game->m_score2 >= 4)
    {
        int minusResult = game->m_score1 - game->m_score2;
        if (minusResult == 1)
            strcpy(game->score, "Advantage player1");
        else if (minusResult == -1)
            strcpy(game->score, "Advantage player2");
        else if (minusResult >= 2)
            strcpy(game->score, "Win for player1");
        else
            strcpy(game->score, "Win for player2");
    }
    else
    {
        for (int i = 1; i < 3; i++)
        {
            if (i == 1) tempScore = game->m_score1;
            else { strcat(game->score, "-"); tempScore = game->m_score2; }
            switch (tempScore)
            {
            case 0:
                strcat(game->score, "Love");
                break;
            case 1:
                strcat(game->score, "Fifteen");
                break;
            case 2:
                strcat(game->score, "Thirty");
                break;
            case 3:
                strcat(game->score, "Forty");
                break;
            }
        }
    }
    return game->score;
}

File: csharp/Tennis/TennisGame1.cs
namespace Tennis
{
    public class TennisGame1 : ITennisGame
    {
        private readonly Player player1;
        private readonly Player player2;

        private Player advantage;
        private Player winner;
        private string scoreDescription;

        public TennisGame1(string player1Name, string player2Name)
        {
            player1 = new Player(player1Name);
            player2 = new Player(player2Name);
        }

        public int NumberOfGamesWon { get; private set; }

        public void WonPoint(string playerName)
        {
            // Note: Comparing the player name against a magic string is prone to
            // encounter errors, as it relies on the player1Name variable passed
            // to the constructor being set to "player1". Requiring this internally
            // will either lead to a bug if the name is not that, or makes the
            // specification of the player name redundant.
            // 
            // An improvement would be:
            //
            // if (playerName == player1.Name)
            if (playerName == "player1")
            {
                player1.IncreaseScore();
            }
            else
            {
                player2.IncreaseScore();
            }
        }

        // Note: Ideally, we would perform this check as soon as a point is won,
        // but since the tests set all of the points upfront and THEN want to
        // check the final state, we need to separate it out into 2 steps. Otherwise,
        // we will trigger a win condition prematurely in the tests.
        public void CheckForWinner()
        {
            CompareScores();
            UpdateScoreDescription();
            ResetScoresIfThereIsAWinner();
        }

        private void CompareScores()
        {
            if (!EitherPlayerScoreIsWinner())
            {
                return;
            }

            // Note: Refactored to using methods instead of magic numbers here
            // to improve readability. It does result in extra mathmatical
            // operations at runtime, but the performance implications will be
            // negligible.
            if (player1.HasAdvantageOver(player2))
            {
                advantage = player1;
            }
            else if (player2.HasAdvantageOver(player1))
            {
                advantage = player2;
            }
            else if (player1.HasWinningScoreAgainst(player2))
            {
                winner = player1;

                NumberOfGamesWon++;
            }
            else
            {
                // Player 2 must have a 2 point lead, and is therefore the winner
                winner = player2;

                NumberOfGamesWon++;
            }
        }

        private void UpdateScoreDescription()
        {
            if (PlayerScoresAreEqual())
            {
                scoreDescription = DetermineEqualScoreState(player1.Score);
            }
            else if (EitherPlayerScoreIsWinner())
            {
                scoreDescription = DeterminePlayerAdvantage();
            }
            else
            {
                scoreDescription = $"{player1.ScoreString}-{player2.ScoreString}";
            }
        }

        private bool PlayerScoresAreEqual() => player1.Score == player2.Score;

        // TODO: Use a meaningful constant value instead of a magic number here.
        private bool EitherPlayerScoreIsWinner() => player1.Score >= 4 || player2.Score >= 4;

        private static string DetermineEqualScoreState(int score)
        {
            // TODO: Use const values instead of magic strings.
            switch (score)
            {
                case 0:
                    return "Love-All";

                case 1:
                    return "Fifteen-All";

                case 2:
                    return "Thirty-All";

                default:
                    return "Deuce";
            }
        }

        private string DeterminePlayerAdvantage()
        {
            if (winner != null)
            {
                return $"Win for {winner.Name}";
            }

            return $"Advantage {advantage.Name}";
        }

        private void ResetScoresIfThereIsAWinner()
        {
            if (winner != null)
            {
                advantage = null;
                winner = null;

                player1.ResetScore();
                player2.ResetScore();
            }
        }

        public string GetScore() => scoreDescription;
    }
}

